std/experimental/checkedint.d

```diff
@@ -0,0 +1,2411 @@
+    /**
+
+    Called automatically upon a bad `opOpAssign` call (one that loses precision
+    or attempts to convert a negative value to an unsigned type).
+
+    Params:
+    rhs = The right-hand side value in the assignment, after the operator has
+    been evaluated
+
+    Returns: Nominally the result is the desired value of the operator, which
+    will be forwarded as result. For `Abort`, the function never returns because
+    it aborts the program.
+
+    */
+    Lhs onBadOpOpAssign(Lhs, Rhs)(Rhs rhs)
+    {
+        Mumble.onBadOpOpAssign!Lhs(rhs);
+        assert(0);
+    }
+
+    /**
+
+    Called automatically upon a bad `opEquals` call (one that would make a
+    signed negative value appear equal to an unsigned positive value).
+
+    Params:
+    lhs = The first argument of `Checked`, e.g. `int` if the left-hand side of
+      the operator is `Checked!int`
+    rhs = The right-hand side type involved in the operator
+
+    Returns: Nominally the result is the desired value of the operator, which
+    will be forwarded as result. For `Abort`, the function never returns because
+    it aborts the program.
+
+    */
+    bool onBadOpEquals(Lhs, Rhs)(Lhs lhs, Rhs rhs)
+    {
+        Mumble.onBadOpEquals(lhs, rhs);
+        assert(0);
+    }
+
+    /**
+
+    Called automatically upon a bad `opCmp` call (one that would make a signed
+    negative value appear greater than or equal to an unsigned positive value).
+
+    Params:
+    lhs = The first argument of `Checked`, e.g. `int` if the left-hand side of
+      the operator is `Checked!int`
+    rhs = The right-hand side type involved in the operator
+
+    Returns: Nominally the result is the desired value of the operator, which
+    will be forwarded as result. For `Abort`, the function never returns because
+    it aborts the program.
+
+    */
+    bool onBadOpCmp(Lhs, Rhs)(Lhs lhs, Rhs rhs)
+    {
+        Mumble.onBadOpCmp(lhs, rhs);
+        assert(0);
+    }
+
+    /**
+
+    Called automatically upon an overflow during a unary or binary operation.
+
+    Params:
+    Lhs = The first argument of `Checked`, e.g. `int` if the left-hand side of
+      the operator is `Checked!int`
+    Rhs = The right-hand side type involved in the operator
+
+    Returns: Nominally the result is the desired value of the operator, which
+    will be forwarded as result. For `Abort`, the function never returns because
+    it aborts the program.
+
+    */
+    typeof(~Lhs()) onOverflow(string x, Lhs)(Lhs lhs)
+    {
+        Mumble.onOverflow!x(lhs);
+        assert(0);
+    }
+    /// ditto
+    typeof(Lhs() + Rhs()) onOverflow(string x, Lhs, Rhs)(Lhs lhs, Rhs rhs)
+    {
+        Mumble.onOverflow!x(lhs, rhs);
+        assert(0);
+    }
+}
+
+unittest
+{
+    Checked!(int, Abort) x;
+    x = 42;
+    short x1 = cast(short) x;
+    //x += long(int.max);
+}
+
+// Mumble
+/**
+Hook that prints to `stderr` a trace of all integral errors, without affecting
+default behavior.
+*/
+struct Mumble
+{
```

Comment:
I know that this is bike-shedding, but I honestly didn't know the word "mumble" before.
Can't we use a common term like "Warn"?